src/build_vehicle_gui.cpp

```diff
@@ -269,20 +269,25 @@ static bool EnginePowerVsRunningCostSorter(const EngineID &a, const EngineID &b)
 {
 	const Engine *e_a = Engine::Get(a);
 	const Engine *e_b = Engine::Get(b);
+	uint p_a = e_a->GetPower();
+	uint p_b = e_b->GetPower();
+	Money r_a = e_a->GetRunningCost();
+	Money r_b = e_b->GetRunningCost();
+	/* Check if running cost is zero in one or both engines */
+	if (r_a == 0) {
+		if (r_b == 0) {
+			if (p_a == p_b) return EngineNumberSorter(a,b);
+			return _engine_sort_direction ? p_a >= p_b : p_a < p_b;
+		}
+		return !_engine_sort_direction;
+	}
+	if (r_b == 0 ) return _engine_sort_direction;
 
-	/* Here we are using a few tricks to get the right sort.
-	 * We want power/running cost, but since we usually got higher running cost than power and we store the result in an int,
-	 * we will actually calculate cunning cost/power (to make it more than 1).
-	 * Because of this, the return value have to be reversed as well and we return b - a instead of a - b.
-	 * Another thing is that both power and running costs should be doubled for multiheaded engines.
-	 * Since it would be multiplying with 2 in both numerator and denominator, it will even themselves out and we skip checking for multiheaded. */
-	Money va = (e_a->GetRunningCost()) / max(1U, (uint)e_a->GetPower());
-	Money vb = (e_b->GetRunningCost()) / max(1U, (uint)e_b->GetPower());
-	int r = ClampToI32(vb - va);
-
+	double va = (double)p_a / (double)r_a;
+	double vb = (double)p_b / (double)r_b;
 	/* Use EngineID to sort instead since we want consistent sorting */
-	if (r == 0) return EngineNumberSorter(a, b);
-	return _engine_sort_direction ? r > 0 : r < 0;
+	if (va == vb) return EngineNumberSorter(a,b);
+	return _engine_sort_direction ? va > vb : va < vb;
```

Comment:
probably unrelated, but i feel like this kind of expression could be abbreviated with `direction XOR comparison`

Reply:
Yeah i thought it could be simplified with some logic, if i put some time to it, but i opted for more clarity.

Reply:
dunno in which mindset a ternary operator is "clearer" than XOR...

Reply:
how would i go about it? i am unsure

Reply:
there's two questions to be answered here:
1) which way around does the comparison go?
2) is this logically correct?

the first question can be answered easily,
given a ternary operator `a?b:c`, where c is false whenever b is true and the other way around, there's two options: `a XOR b` or `a XOR c`

a|a?b:c|a XOR b|a XOR c
-|-|-|-
TRUE|TRUE/FALSE|FALSE/TRUE|TRUE/FALSE
FALSE|FALSE/TRUE|TRUE/FALSE|FALSE/TRUE

as we see, `a XOR c` is the right way around

now for the second question, the logical correctness.

the problem here is that C/C++ to my knowledge only offers a bitwise XOR, not a logical XOR, so to ensure correctness here, a,b and c must be guaranteed to only contain the values `0` and `1`. however, i don't do enough C/C++ to know if that is the case.

Reply:
I have been thinking more and i am not sure if using ^ or != is more correct.

Reply:
Don't do stupid xor tricks here, the decrease in readability is not worth it.

Reply:
Yeah. I opted for using !=.